src/validity.cpp

```diff
@@ -1219,68 +1252,77 @@ bool is_label_valid(const std::u32string_view label) {
    //       Bidi property NSM.
 
 
-  auto find_direction = [](uint32_t code) -> direction {
-    auto it = std::lower_bound(std::begin(dir_table), std::end(dir_table), code, 
-    [](const directions& d, uint32_t c) { return d.final_code < c; });
-    // next check is almost surely in vain, but we use it for safety.
-    if(it == std::end(dir_table)) { return direction::NONE; }
-    // We have that d.final_code >= c.
-    if (code >= it->start_code) { return it->direct; }
-    return direction::NONE;
-  };
-  bool have_bidi = false;
-  for (size_t i = 0; i < label.size(); i++) {
-    auto d = find_direction(label[i]);
-    if ((d == direction::R) || (d == direction::AL) || (d == direction::AN)) {
-      have_bidi = true;
-    }
+  bool is_bidi_domain = is_rtl_label(label);
+  size_t last_non_nsm_char = find_last_not_of_nsm(label);
+  if(last_non_nsm_char ==  std::u32string_view::npos) {
+    return false;
   }
-  if (have_bidi) {
+
+  if(is_bidi_domain) {
     const direction first_d = find_direction(label[0]);
-    if (!(first_d == direction::L || first_d == direction::R ||
-          first_d == direction::AL)) {
-      return false;
-    }
-    bool rtl = (first_d == direction::R || first_d == direction::AL);
-    bool ve = false;
-    direction nt = direction::NONE;
-    for (size_t i = 1; i < label.size(); i++) {
-      const auto d = find_direction(label[i]);
-      if (rtl) {
-        if (!(d == direction::R || d == direction::AL || d == direction::AN ||
+
+    // The first character must be a character with Bidi property L, R,
+    // or AL.  If it has the R or AL property, it is an RTL label; if it
+    // has the L property, it is an LTR label.
+    bool eval_as_rtl = (first_d == direction::AL) || (first_d == direction::R);
```

Comment:
I'm not so sure.. 
In fact "An LTR label is any label that is not an RTL label.", 
But when it comes to how it should be evaluated, the specs says:
"The first character must be a character with Bidi property L, R, or AL.  If it has the R or AL property, it is an RTL label; if it has the L property, it is an LTR label."  (the first character kind of defines how we should look into the label (thats how I undertand it) )
So given the conditions that states how we should look into the label based on the first char, it don`t seems like we could simply do something like:
```c++
bool eval_as_rtl = condition;
bool eval_as_ltr = !eval_as_rtl;
```
Saying this here but not even tested yet to see if it would still pass the tests :sweat_smile:

Reply:
Following your suggestion (it passed the tests)